journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java

```diff
@@ -105,72 +133,97 @@ public void append(final JournalRecord record) {
+      buffer.position(startPosition);
+      throw boe;
+    }
 
-    writeInternal(buffer, recordToWrite);
-    return recordToWrite;
-  }
+    final long checksum = computeChecksum(buffer, startPosition + metadataLength, recordLength);
 
-  /**
-   * Write the record to the buffer. After the method returns, the position of buffer will be
-   * advanced to a position were the next record will be written.
-   */
-  private JournalRecord write(final ByteBuffer buffer, final JournalRecord record) {
-    final var checksum = checksumGenerator.compute(record.data());
-    if (checksum != record.checksum()) {
-      throw new InvalidChecksum("Checksum invalid for record " + record);
+    if (record.checksum() != checksum) {
+      buffer.position(startPosition);
+      throw new InvalidChecksum(
+          String.format("Failed to append record %s. Checksum does not match", record));
     }
-    writeInternal(buffer, record);
-    return record;
+
+    final JournalRecordMetadata metadata =
+        writeMetadata(buffer, writeBuffer, startPosition, metadataLength, recordLength, checksum);
+
+    updateLastWrittenEntry(metadata, indexedRecord, startPosition);
+
+    buffer.position(startPosition + metadataLength + recordLength);
   }
 
-  private void writeInternal(final ByteBuffer buffer, final JournalRecord recordToWrite) {
-    final int recordStartPosition = buffer.position();
-    buffer.mark();
-    if (recordStartPosition + Integer.BYTES > buffer.limit()) {
-      throw new BufferOverflowException();
-    }
+  private void updateLastWrittenEntry(
+      final JournalRecordMetadata metadata,
+      final JournalIndexedRecord recordRead,
+      final int startPosition) {
+    lastEntry = new PersistedJournalRecord(metadata, recordRead);
+    index.index(lastEntry, startPosition);
+  }
 
-    buffer.position(recordStartPosition + Integer.BYTES);
-    try {
-      serializer.write(recordToWrite, buffer);
-    } catch (final BufferOverflowException e) {
-      buffer.reset();
-      throw e;
-    }
+  private JournalRecordMetadata writeMetadata(
+      final ByteBuffer buffer,
+      final MutableDirectBuffer writeBuffer,
+      final int startPosition,
+      final int metadataLength,
+      final int recordLength,
+      final long checksum) {
+    writeBuffer.wrap(buffer, startPosition, metadataLength);
+    final JournalRecordMetadata recordMetadata =
+        new JournalRecordMetadataImpl(checksum, recordLength);
+    serializer.write(recordMetadata, writeBuffer);
+    return recordMetadata;
+  }
 
-    final int length = buffer.position() - (recordStartPosition + Integer.BYTES);
+  private int writeRecord(
+      final ByteBuffer buffer,
+      final MutableDirectBuffer writeBuffer,
+      final int offset,
+      final JournalIndexedRecord indexedRecord) {
+    writeBuffer.wrap(buffer, offset, buffer.limit() - offset);
+    final var recordLength = serializer.write(indexedRecord, writeBuffer);
+    return recordLength;
```

Comment:
Can just return immediately here

```suggestion
    return serializer.write(indexedRecord, writeBuffer);
```